### PR TITLE
Implement SASL authentication

### PR DESCRIPTION
**Pull request by spkrka**

### CODE REVIEW

**Review by mattnworb — COMMENTED** [review]
This is an unsolicited review, but I was curious how authentication in memcache worked in general so I took a look, and code LGTM 👍

File: folsom/src/main/java/com/spotify/folsom/authenticate/AsciiAuthenticationValidator.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+                throw new CompletionException(unwrap(throwable));
+              }
+            }))
+    );
+  }
+
+  @Override
+  public void validate(final boolean binary) {
+    if (binary) {
+      throw new RuntimeException("Programmer error: wrong validator used");
```
[the remark this answers is not in the document]
Comment (spkrka) [reply folsom/src/main/java/com/spotify/folsom/authenticate/AsciiAuthenticationValidator.java]: Good point, will fix.

File: folsom/src/main/java/com/spotify/folsom/authenticate/AsciiAuthenticationValidator.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import com.spotify.folsom.client.ascii.GetRequest;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CompletionStage;
+
+/**
+ * If SASL is enabled than ascii protocol won't be allowed at all, so we simply expect a connection closed.
+ */
+public class AsciiAuthenticationValidator implements Authenticator {
+
+  public static final AsciiAuthenticationValidator INSTANCE
```
[the remark this answers is not in the document]
Comment (spkrka) [reply folsom/src/main/java/com/spotify/folsom/authenticate/AsciiAuthenticationValidator.java]: I agree in general, but this should be considered internal. I would have made it non-public if it wasn't in separate sub-packages.

**Comment by spkrka** [comment]
Unsolicited or not, thanks for the review @mattnworb !

**Comment by bcleenders** [comment]
👍 from me (but should get a non-comitter review, I guess)

File: folsom/src/main/java/com/spotify/folsom/MemcacheClientBuilder.java [hunk 3]
```diff
@@ -406,22 +412,51 @@ public MemcacheClientBuilder(final Transcoder<V> valueTranscoder) {
...
+    return this;
+  }
+
+
   /**
    * Create a client that uses the binary memcache protocol.
    * @return a {@link com.spotify.folsom.BinaryMemcacheClient}
    */
   public BinaryMemcacheClient<V> connectBinary() {
+    final Authenticator authenticator = firstNonNull(
```
Comment (protocol7) [inline folsom/src/main/java/com/spotify/folsom/MemcacheClientBuilder.java]: Having a potentially null field seems unnecessary. How about initializing it with BinaryAuthenticationValidator.INSTANCE?

File: folsom/src/main/java/com/spotify/folsom/authenticate/AsciiAuthenticationValidator.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import static com.spotify.folsom.client.Utils.unwrap;
+
+import com.google.common.base.Charsets;
+import com.spotify.folsom.RawMemcacheClient;
+import com.spotify.folsom.client.ascii.GetRequest;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CompletionStage;
+
+/**
+ * If SASL is enabled than ascii protocol won't be allowed at all, so we simply expect a connection closed.
```
Comment (protocol7) [inline folsom/src/main/java/com/spotify/folsom/authenticate/AsciiAuthenticationValidator.java]: s/than/then/

File: folsom/src/main/java/com/spotify/folsom/authenticate/AsciiAuthenticationValidator.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+ */
+public class AsciiAuthenticationValidator implements Authenticator {
+
+  public static final AsciiAuthenticationValidator INSTANCE
+      = new AsciiAuthenticationValidator();
+
+  private static final byte[] EXAMPLE_KEY = "a".getBytes(Charsets.US_ASCII);
```
Comment (protocol7) [inline folsom/src/main/java/com/spotify/folsom/authenticate/AsciiAuthenticationValidator.java]: Did we want to switch this to something that is obvious if one would look at the wire, e.g. "authentication_check" or similar?

File: folsom/src/main/java/com/spotify/folsom/authenticate/AsciiAuthenticationValidator.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+  private AsciiAuthenticationValidator() {
+  }
+
+  @Override
+  public CompletionStage<RawMemcacheClient> authenticate(
+      CompletionStage<RawMemcacheClient> clientFuture) {
```
Comment (protocol7) [inline folsom/src/main/java/com/spotify/folsom/authenticate/AsciiAuthenticationValidator.java:46]: Missing final

File: folsom/src/main/java/com/spotify/folsom/authenticate/AuthenticatingClient.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+package com.spotify.folsom.authenticate;
+
+import com.spotify.folsom.RawMemcacheClient;
+import com.spotify.folsom.reconnect.Connector;
+import java.util.concurrent.CompletionStage;
+
+public class AuthenticatingClient {
+
+  public static CompletionStage<RawMemcacheClient> authenticate(
+      Connector connector,
```
Comment (protocol7) [inline folsom/src/main/java/com/spotify/folsom/authenticate/AuthenticatingClient.java:26]: Missing final

File: folsom/src/main/java/com/spotify/folsom/authenticate/PlaintextAuthenticator.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+import com.spotify.folsom.client.binary.PlaintextAuthenticateRequest;
+import java.util.concurrent.CompletionStage;
+
+public class PlaintextAuthenticator implements Authenticator {
+
+  private final String username;
+  private final String password;
+
+  public PlaintextAuthenticator(String username, String password) {
+    this.username = username;
```
Comment (protocol7) [inline folsom/src/main/java/com/spotify/folsom/authenticate/PlaintextAuthenticator.java]: Null check parameters?

File: folsom/src/main/java/com/spotify/folsom/authenticate/PlaintextAuthenticator.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+import com.spotify.folsom.RawMemcacheClient;
+import com.spotify.folsom.client.binary.PlaintextAuthenticateRequest;
+import java.util.concurrent.CompletionStage;
+
+public class PlaintextAuthenticator implements Authenticator {
+
+  private final String username;
+  private final String password;
+
+  public PlaintextAuthenticator(String username, String password) {
```
Comment (protocol7) [inline folsom/src/main/java/com/spotify/folsom/authenticate/PlaintextAuthenticator.java]: Missing final

File: folsom/src/main/java/com/spotify/folsom/authenticate/PlaintextAuthenticator.java [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    this.username = username;
+    this.password = password;
+  }
+
+  public CompletionStage<RawMemcacheClient> authenticate(
+      CompletionStage<RawMemcacheClient> clientFuture) {
```
Comment (protocol7) [inline folsom/src/main/java/com/spotify/folsom/authenticate/PlaintextAuthenticator.java]: Final

File: folsom/src/main/java/com/spotify/folsom/authenticate/PlaintextAuthenticator.java [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+    PlaintextAuthenticateRequest
```
Comment (protocol7) [inline folsom/src/main/java/com/spotify/folsom/authenticate/PlaintextAuthenticator.java]: Final

File: folsom/src/main/java/com/spotify/folsom/client/binary/PlaintextAuthenticateRequest.java [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+  private static final byte[] KEY = "PLAIN".getBytes(Charsets.US_ASCII);
+
+  private final String username;
+  private final String password;
+
+  public PlaintextAuthenticateRequest(final String username, final String password) {
+    // Key is auth type
+    super(KEY);
+
+    this.username = username;
```
Comment (protocol7) [inline folsom/src/main/java/com/spotify/folsom/client/binary/PlaintextAuthenticateRequest.java]: Null check?

File: folsom/src/main/java/com/spotify/folsom/reconnect/ReconnectingClient.java [hunk 13]
```diff
@@ -48,18 +52,20 @@
 
   private final BackoffFunction backoffFunction;
   private final ScheduledExecutorService scheduledExecutorService;
-  private final Connector connector;
+  private final com.spotify.folsom.reconnect.Connector connector;
```
Comment (protocol7) [inline folsom/src/main/java/com/spotify/folsom/reconnect/ReconnectingClient.java:55]: Is this needed?

File: folsom/src/main/java/com/spotify/folsom/MemcacheClientBuilder.java [hunk 14]
```diff
@@ -406,22 +412,51 @@ public MemcacheClientBuilder(final Transcoder<V> valueTranscoder) {
     return this;
   }
 
+  /**
+   * Authenticate with memcached using plaintest SASL. This only works for binary connections, not ascii.
```
[the remark this answers is not in the document]
Comment (spkrka) [reply folsom/src/main/java/com/spotify/folsom/MemcacheClientBuilder.java]: Will fix!

File: folsom/src/main/java/com/spotify/folsom/MemcacheClientBuilder.java (hunk 3 above)
Comment (spkrka) [reply folsom/src/main/java/com/spotify/folsom/MemcacheClientBuilder.java]: Then the ascii case gets a lot more complicated

File: folsom/src/main/java/com/spotify/folsom/authenticate/AsciiAuthenticationValidator.java (hunk 4 above)
Comment (spkrka) [reply folsom/src/main/java/com/spotify/folsom/authenticate/AsciiAuthenticationValidator.java]: Will fix!

File: folsom/src/main/java/com/spotify/folsom/authenticate/AsciiAuthenticationValidator.java (hunk 5 above)
Comment (spkrka) [reply folsom/src/main/java/com/spotify/folsom/authenticate/AsciiAuthenticationValidator.java]: Will fix!

File: folsom/src/main/java/com/spotify/folsom/authenticate/PlaintextAuthenticator.java (hunk 9 above)
Comment (spkrka) [reply folsom/src/main/java/com/spotify/folsom/authenticate/PlaintextAuthenticator.java]: Will fix!

File: folsom/src/main/java/com/spotify/folsom/authenticate/PlaintextAuthenticator.java (hunk 10 above)
Comment (spkrka) [reply folsom/src/main/java/com/spotify/folsom/authenticate/PlaintextAuthenticator.java]: Will fix!

File: folsom/src/main/java/com/spotify/folsom/authenticate/PlaintextAuthenticator.java (hunk 11 above)
Comment (spkrka) [reply folsom/src/main/java/com/spotify/folsom/authenticate/PlaintextAuthenticator.java]: Will fix!

File: folsom/src/main/java/com/spotify/folsom/reconnect/ReconnectingClient.java (hunk 13 above)
Comment (spkrka) [reply folsom/src/main/java/com/spotify/folsom/reconnect/ReconnectingClient.java:55]: Yes, it's used in the retry method

Comment (protocol7) [reply folsom/src/main/java/com/spotify/folsom/reconnect/ReconnectingClient.java:55]: Sorry, I meant to fully qualified name

Comment (spkrka) [reply folsom/src/main/java/com/spotify/folsom/reconnect/ReconnectingClient.java:55]: Yes, there are two Connector classes (I guess we could rename one of them instead)

File: folsom/src/main/java/com/spotify/folsom/ObservableClient.java [hunk 15]
```diff
@@ -86,4 +128,16 @@ default void awaitFuture(final CompletionStage<Void> future, final long waitTime
    * @return null, if there's no connection failure
    */
   Throwable getConnectionFailure();
+
```
[the remark this answers is not in the document]
Comment (spkrka) [reply folsom/src/main/java/com/spotify/folsom/ObservableClient.java:131]: Correct - I'm usually more strict about that but my reasoning is along what you said.

**Pull request merged**